File: cli/te/models/post.py
from dataclasses import dataclass
from te.models.document import Document
from te.utilities import format_as_slug

import frontmatter


@dataclass
class Post(Document):
    slug: str
    title: str
    tags: list[str]
    is_published: bool
    preview: str
    markdown: str

    directory: str = "posts"

    def to_markdown(self) -> str:
        return f"""---
title: {self.title}
tags: {", ".join(self.tags)}
is_published: {self.is_published}
preview: {self.preview}
---

{self.markdown}"""

    def to_json(self) -> dict:
        ...

    @staticmethod
    def from_file(body: str) -> "Post":
        post: frontmatter.Post = frontmatter.loads(body)

        title = post["title"]
        slug = format_as_slug(title)
        tags = post["tags"]
        is_published = post["is_published"]
        preview = post["preview"]

        markdown = post.content

        return Post(slug, title, tags, is_published, preview, markdown)

    @staticmethod
    def from_json(post: dict) -> "Post":

        return Post(id=post["id"])

File: site/interfaces/Post.ts
/** `Post` interface
 *
 * Expected structure of posts
 *
 * Reflected in `api/models/Post.ts` for db interface
 */
interface Post {
    /** Unique GUID identifying the post */
    _id: string;

    /** Defines the URL path to the post
     *
     * Derived from the title as all lowercase with dashes replacing spaces */
    slug: string;

    /** Title of the post */
    title: string;

    /** Date when the post was created */
    created: Date;

    /** Date when the post was last changed */
    updated: Date;

    /** Tags for categorizing the post (e.g. Python, Programming, Science) */
    tags: string[];

    /** Whether the post is listed on the homepage */
    isPublished: boolean;

    /** Text describing the post */
    preview: string;

    /** Body of the post, formatted in markdown */
    markdown: string;
}

export default Post;
